src/lib/Type.hs

```diff
@@ -115,9 +115,29 @@ getClassTy (ClassDef _ _ bs _ _) = go bs
 
 -- === querying effects ===
 
-exprEffects :: (MonadFail1 m, EnvReader m) => Expr n -> m n (EffectRow n)
+computeAbsEffects :: (EnvExtender m, SubstE Name e)
+  => Abs (Nest Decl) e n -> m n (Abs (Nest Decl) (EffectRow `PairE` e) n)
+computeAbsEffects it = refreshAbs it \decls result -> do
+  effs <- declNestEffects decls
+  return $ Abs decls (effs `PairE` result)
+
+declNestEffects :: (EnvReader m) => Nest Decl n l -> m l (EffectRow l)
+declNestEffects decls = liftEnvReaderM $ declNestEffectsRec decls
```

Comment:
Done.  Also inlined `withType` in Builder.hs for the same reason, and `makeBlock` because it was already inlined before (though it's pure, so maybe it doesn't matter as much).